### PR TITLE
feat(Collector): add lastCollectedTimestamp

### PR DESCRIPTION
**Pull request by ImRodry**
**Please describe the changes this PR makes and why it should be merged:**

This PR adds a new property and a new getter to the Collector class to store the timestamp at which the collector last collected an item. This is useful for determining when the collector will end when an idle timeout is specified as there is currently no reliable way to get this information from Node without using private properties of the Timeout class and weird hacks

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

**Comment by jaw0r3k** [comment]
you should chanhe the name

**Comment by ImRodry** [comment]
The formatting errors are due to an update to prettier that changed a rule. Should I commit those changes?

File: packages/discord.js/src/structures/interfaces/Collector.js [hunk 1]
```diff
@@ -97,6 +97,20 @@ class Collector extends EventEmitter {
...
 
+  /**
+   * The timestamp at which this collector last collected an item
+   * @type {number|null}
+   */
+  lastCollectedTimestamp = null;
+
+  /**
+   * The Date at which this collector last collected an item
+   * @type {Date|null}
```
Comment (jaw0r3k) [inline packages/discord.js/src/structures/interfaces/Collector.js:108]:
```suggestion
   * @type {?Date}
```

File: packages/discord.js/src/structures/interfaces/Collector.js [hunk 2]
```diff
@@ -97,6 +97,20 @@ class Collector extends EventEmitter {
     if (options.idle) this._idletimeout = setTimeout(() => this.stop('idle'), options.idle).unref();
   }
 
+  /**
+   * The timestamp at which this collector last collected an item
+   * @type {number|null}
```
Comment (Syjalo) [inline packages/discord.js/src/structures/interfaces/Collector.js:102]:
```suggestion
   * @type {?number}
```

File: packages/discord.js/src/structures/interfaces/Collector.js (hunk 1 above)
Comment (Syjalo) [inline packages/discord.js/src/structures/interfaces/Collector.js:108]:
```suggestion
   * @type {?Date}
```

**Review by Syjalo — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/interfaces/Collector.js (hunk 2 above)
Comment (ImRodry) [reply packages/discord.js/src/structures/interfaces/Collector.js:102]: Some properties on this class are typed as |null others with ? I’m not sure which one to use but |null seems more appropriate

Comment (jaw0r3k) [reply packages/discord.js/src/structures/interfaces/Collector.js:102]: Yeah also for me |null is more clear

Comment (Syjalo) [reply packages/discord.js/src/structures/interfaces/Collector.js:102]: `|null` is used only in 2 places. In all other places `?` is used.

https://github.com/KhafraDev/discord.js/blob/5befbc3773a1cf33614bf900f8301686130b33c7/packages/discord.js/src/structures/interfaces/Collector.js#L83

https://github.com/ckohen/discord.js/blob/8dfd0037b53ec6ef0900a53a0fb1b68fa2cbaf72/packages/discord.js/src/structures/ApplicationCommand.js#L147

Comment (ImRodry) [reply packages/discord.js/src/structures/interfaces/Collector.js:102]: Exactly, it’s used in this file. I think `?` is for optional properties which this one isn’t

Comment (kyranet) [reply packages/discord.js/src/structures/interfaces/Collector.js:102]: `?number` and `number|null` are equivalent, the `?` doesn't denote optionality (like in TypeScript), but nullability:

image

File: packages/discord.js/src/structures/interfaces/Collector.js [hunk 3]
```diff
@@ -97,6 +97,20 @@ class Collector extends EventEmitter {
     if (options.idle) this._idletimeout = setTimeout(() => this.stop('idle'), options.idle).unref();
   }
 
+  /**
+   * The timestamp at which this collector last collected an item
+   * @type {?number}
+   */
+  lastCollectedTimestamp = null;
```
Comment (kyranet) [inline packages/discord.js/src/structures/interfaces/Collector.js:104]: All the properties are assigned in the constructor, not as fields, this should be moved inside the constructor.

**Review by kyranet — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/interfaces/Collector.js (hunk 3 above)
Comment (ImRodry) [reply packages/discord.js/src/structures/interfaces/Collector.js:104]: don't really get why but I'll change this

**Comment by ImRodry** [comment]
would love to know how in the world I removed those review requests lol

**Review by kyranet — APPROVED** [review]

**Comment by Syjalo** [comment]
Probably you can also add `endingAt`

**Comment by ImRodry** [comment]
maybe a new pr

**Review by SpaceEEC — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**